From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Thu, 10 Jul 2025 12:16:28 +0700
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index afaff6dcfe2044c465693dd0971d07644470ba8a..1f3db338ccd971547a9d0cc1a6329de3f7d01ac4 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -37,6 +37,7 @@ public record ServerBuildInfoImpl(
     private static final String BRAND_PURPUR_NAME = "Purpur";
     private static final String BRAND_LEAF_NAME = "Leaf";
     // Leaf end - Rebrand & Fix Pufferfish and Purpur patches
+    private static final String BRAND_LEAF_CWS = "Leaf-CWS";
 
     private static final String BUILD_DEV = "DEV";
 
@@ -48,9 +49,9 @@ public record ServerBuildInfoImpl(
         this(
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_ID)
                 .map(Key::key)
-                .orElse(BRAND_LEAF_ID), // Gale - branding changes // Leaf - Rebrand
+                .orElse(BRAND_LEAF_CWS_ID), // Gale - branding changes // Leaf - Rebrand
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_NAME)
-                .orElse(BRAND_LEAF_NAME), // Gale - branding changes // Leaf - Rebrand
+                .orElse(BRAND_LEAF_CWS), // Gale - branding changes // DomamaN202 - Rebrand
             SharedConstants.getCurrentVersion().id(),
             SharedConstants.getCurrentVersion().name(),
             getManifestAttribute(manifest, ATTRIBUTE_BUILD_NUMBER)
@@ -72,7 +73,8 @@ public record ServerBuildInfoImpl(
             || brandId.equals(BRAND_PAPER_ID)
             || brandId.equals(BRAND_GALE_ID) // Gale - branding changes
             || brandId.equals(BRAND_PUFFERFISH_ID)
-            || brandId.equals(BRAND_PURPUR_ID);
+            || brandId.equals(BRAND_PURPUR_ID)
+            || brandId.equals(BRAND_LEAF_ID); // DomamaN202 - branding changes
             // Leaf end - Rebrand & Fix Pufferfish and Purpur patches
     }
 
