From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Fri, 11 Jul 2025 08:45:46 +0700
Subject: [PATCH] CWS Main.onSoundEventsInit Impl


diff --git a/net/minecraft/sounds/SoundEvents.java b/net/minecraft/sounds/SoundEvents.java
index f08f8428db65eb5a462fd625a8f8c160dc37cc8f..fcb6b59b0f299206d17748d8b7c8ac5460d45d29 100644
--- a/net/minecraft/sounds/SoundEvents.java
+++ b/net/minecraft/sounds/SoundEvents.java
@@ -5,6 +5,7 @@ import java.util.Map;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
+
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
 import net.minecraft.core.registries.BuiltInRegistries;
@@ -1746,4 +1747,10 @@ public class SoundEvents {
                 )
             );
     }
+
+    // DomamaN202 start - Register Sounds
+    static {
+        ru.cws.Main.onSoundEventsInit();
+    }
+    // DomamaN202 end - Register Sounds
 }
