From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Fri, 11 Jul 2025 11:31:39 +0700
Subject: [PATCH] CWS-Impl (BetterCombat)


diff --git a/net/minecraft/core/component/DataComponents.java b/net/minecraft/core/component/DataComponents.java
index 3c97531a7bbcd84b80d4704aaa27b913652f4019..785b8adf3c3f799564c47504e28e9de0d73fc360 100644
--- a/net/minecraft/core/component/DataComponents.java
+++ b/net/minecraft/core/component/DataComponents.java
@@ -393,4 +393,10 @@ public class DataComponents {
     private static <T> DataComponentType<T> register(String name, UnaryOperator<DataComponentType.Builder<T>> builder) {
         return Registry.register(BuiltInRegistries.DATA_COMPONENT_TYPE, name, builder.apply(DataComponentType.builder()).build());
     }
+
+    // CWS start - (BetterCombat) Init BetterCombatDataComponents class
+    static {
+        var abc = net.bettercombat.api.component.BetterCombatDataComponents.WEAPON_PRESET_ID;  // Static class init tail inject
+    }
+    // CWS end - (BetterCombat) Init BetterCombatDataComponents class
 }
diff --git a/net/minecraft/resources/ResourceLocation.java b/net/minecraft/resources/ResourceLocation.java
index 98e27b3cea8ee8c0ba3445a3153bc228c104f3df..58b71c605cdd07de8e84ac3e9ac49a6b7c4397c2 100644
--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -24,6 +24,7 @@ public final class ResourceLocation implements Comparable<ResourceLocation> {
     public static final String DEFAULT_NAMESPACE = "minecraft";
     public static final String REALMS_NAMESPACE = "realms";
     public static final String PAPER_NAMESPACE = "paper"; // Paper
+    public static final String BETTERCOMBAT_NAMESPACE = "bettercombat"; // CWS - (BetterCombat) Add BetterCombat ID to resource list
     private final String namespace;
     private final String path;
 
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index a16607c165965a7c325909ded75a3cf65440f6f7..ddbd62ec2c99659203845857ed063246693a976f 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -3,6 +3,8 @@ package net.minecraft.server.level;
 import com.google.common.cache.CacheBuilder;
 import com.google.common.cache.CacheLoader;
 import com.google.common.cache.LoadingCache;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
 import com.google.common.hash.HashCode;
 import com.google.common.net.InetAddresses;
 import com.mojang.authlib.GameProfile;
@@ -42,6 +44,7 @@ import net.minecraft.core.component.TypedDataComponent;
 import net.minecraft.core.particles.BlockParticleOption;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.PacketSendListener;
 import net.minecraft.network.chat.ChatType;
 import net.minecraft.network.chat.CommonComponents;
@@ -65,7 +68,6 @@ import net.minecraft.network.protocol.game.ClientboundHorseScreenOpenPacket;
 import net.minecraft.network.protocol.game.ClientboundHurtAnimationPacket;
 import net.minecraft.network.protocol.game.ClientboundMerchantOffersPacket;
 import net.minecraft.network.protocol.game.ClientboundOpenBookPacket;
-import net.minecraft.network.protocol.game.ClientboundOpenScreenPacket;
 import net.minecraft.network.protocol.game.ClientboundOpenSignEditorPacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerAbilitiesPacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerCombatEndPacket;
@@ -86,6 +88,9 @@ import net.minecraft.network.protocol.game.ClientboundSystemChatPacket;
 import net.minecraft.network.protocol.game.ClientboundUpdateMobEffectPacket;
 import net.minecraft.network.protocol.game.CommonPlayerSpawnInfo;
 import net.minecraft.network.protocol.status.ServerStatus;
+import net.minecraft.network.syncher.EntityDataAccessor;
+import net.minecraft.network.syncher.EntityDataSerializers;
+import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
@@ -101,10 +106,8 @@ import net.minecraft.stats.ServerRecipeBook;
 import net.minecraft.stats.ServerStatsCounter;
 import net.minecraft.stats.Stat;
 import net.minecraft.stats.Stats;
-import net.minecraft.tags.FluidTags;
 import net.minecraft.util.HashOps;
 import net.minecraft.util.Mth;
-import net.minecraft.util.RandomSource;
 import net.minecraft.util.Unit;
 import net.minecraft.world.Container;
 import net.minecraft.world.Difficulty;
@@ -191,7 +194,7 @@ import net.minecraft.world.scores.Team;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 import org.slf4j.Logger;
 
-public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patches.chunk_system.player.ChunkSystemServerPlayer { // Paper - rewrite chunk system
+public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patches.chunk_system.player.ChunkSystemServerPlayer, net.bettercombat.logic.PlayerAttackProperties { // Paper - rewrite chunk system // CWS - (BetterCombat) Combo system impl
     private static final Logger LOGGER = LogUtils.getLogger();
     public long lastSave = MinecraftServer.currentTick; // Paper - Incremental chunk and player saving
     private static final int NEUTRAL_MOB_DEATH_NOTIFICATION_RADII_XZ = 32;
@@ -809,6 +812,13 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         }
         // Purpur end - Ridables
         org.dreeam.leaf.protocol.Protocols.tickPlayer(this); // Leaf - Protocol core
+
+        // CWS start - (BetterCombat) Player animation
+        var pose = net.bettercombat.logic.PlayerAttackHelper.poseForPlayer(this);
+        this.entityData.set(BETTER_COMBAT_MAIN_IDLE_ANIMATION, pose.base());
+        this.entityData.set(BETTER_COMBAT_OFF_IDLE_ANIMATION, pose.offHand());
+        updateDualWieldingSpeedBoost();
+        // CWS end - (BetterCombat) Player animation
     }
 
     private void updatePlayerAttributes() {
@@ -3208,4 +3218,157 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.ramBar = ramBar;
     }
     // Purpur end - Add rambar command
+
+    // CWS start - (BetterCombat)
+    private int comboCount = 0;
+
+    public int getComboCount() {
+        return comboCount;
+    }
+
+    public void setComboCount(int comboCount) {
+        this.comboCount = comboCount;
+    }
+
+    private static final EntityDataAccessor<String> BETTER_COMBAT_MAIN_IDLE_ANIMATION = SynchedEntityData.defineId(Player.class, EntityDataSerializers.STRING);
+    private static final EntityDataAccessor<String> BETTER_COMBAT_OFF_IDLE_ANIMATION = SynchedEntityData.defineId(Player.class, EntityDataSerializers.STRING);
+
+    @Override
+    protected void defineSynchedData(SynchedEntityData.Builder builder) {
+        super.defineSynchedData(builder);
+        builder.define(BETTER_COMBAT_MAIN_IDLE_ANIMATION, "");
+        builder.define(BETTER_COMBAT_OFF_IDLE_ANIMATION, "");
+    }
+
+    @Override
+    public double blockInteractionRange() {
+        return net.bettercombat.logic.PlayerAttackHelper.getRangeWithWeapon(this, this.getAttributeValue(Attributes.BLOCK_INTERACTION_RANGE));// DomamaN202 - Mixin PlayerEntityRangeMixin
+    }
+
+    private Multimap<Holder<Attribute>, AttributeModifier> dualWieldingAttributeMap;
+    private static final ResourceLocation dualWieldingSpeedModifierId = ResourceLocation.fromNamespaceAndPath(net.bettercombat.BetterCombatMod.ID, "dual_wield");
+
+    private void updateDualWieldingSpeedBoost() {
+        var newState = net.bettercombat.logic.PlayerAttackHelper.isDualWielding(this);
+        var currentState = dualWieldingAttributeMap != null;
+        if (newState != currentState) {
+            if(newState) {
+                // Just started dual wielding
+                // Adding speed boost modifier
+                this.dualWieldingAttributeMap = HashMultimap.create();
+                double multiplier = net.bettercombat.BetterCombatMod.config.server.dual_wielding_attack_speed_multiplier - 1;
+                dualWieldingAttributeMap.put(
+                    Attributes.ATTACK_SPEED,
+                    new AttributeModifier(
+                        dualWieldingSpeedModifierId,
+                        multiplier,
+                        AttributeModifier.Operation.ADD_MULTIPLIED_BASE));
+                this.getAttributes().addTransientAttributeModifiers(dualWieldingAttributeMap);
+            } else {
+                // Just stopped dual wielding
+                // Removing speed boost modifier
+                if (dualWieldingAttributeMap != null) { // Safety first... Who knows...
+                    this.getAttributes().removeAttributeModifiers(dualWieldingAttributeMap);
+                    dualWieldingAttributeMap = null;
+                }
+            }
+        }
+    }
+
+//    @Override
+//    public ItemStack getItemBySlot(EquipmentSlot slot) {
+//        var mainHandHasTwoHanded = false;
+//        var mainHandStack = super.getItemBySlot(EquipmentSlot.MAINHAND);
+//        var mainHandAttributes = net.bettercombat.logic.WeaponRegistry.getAttributes(mainHandStack);
+//        if (mainHandAttributes != null && mainHandAttributes.isTwoHanded()) {
+//            mainHandHasTwoHanded = true;
+//        }
+//
+//        var offHandHasTwoHanded = false;
+//        var offHandStack = super.getItemBySlot(EquipmentSlot.OFFHAND);
+//        var offHandAttributes = net.bettercombat.logic.WeaponRegistry.getAttributes(offHandStack);
+//        if(offHandAttributes != null && offHandAttributes.isTwoHanded()) {
+//            offHandHasTwoHanded = true;
+//        }
+//
+//        if (slot == EquipmentSlot.OFFHAND) {
+//            if (mainHandHasTwoHanded || offHandHasTwoHanded) {
+//                return (ItemStack.EMPTY);
+//            }
+//        }
+//
+//        return super.getItemBySlot(slot);
+//    }
+
+    @Override
+    public InteractionHand getHand$BetterCombat() {
+        var currentHand = net.bettercombat.logic.PlayerAttackHelper.getCurrentAttack(this, comboCount);
+        if (currentHand != null) {
+            return currentHand.isOffHand() ? InteractionHand.OFF_HAND : InteractionHand.MAIN_HAND;
+        } else {
+            return InteractionHand.MAIN_HAND;
+        }
+    }
+
+    private net.bettercombat.api.AttackHand lastAttack;
+
+    @Override
+    public ItemStack getMainHandStack$BetterCombat() {
+        // DUAL WIELDING LOGIC
+        // Here we return the off-hand stack as fake main-hand, purpose:
+        // - Getting enchants
+        // - Getting itemstack to be damaged
+        if (comboCount < 0) {
+            // Vanilla behaviour
+            return this.getMainHandItem();
+        }
+        var hand = net.bettercombat.logic.PlayerAttackHelper.getCurrentAttack(this, comboCount);
+        if (hand == null) {
+            var isOffHand = net.bettercombat.logic.PlayerAttackHelper.shouldAttackWithOffHand(this, comboCount);
+            if (isOffHand) {
+                return ItemStack.EMPTY;
+            } else {
+                return this.getMainHandItem();
+            }
+        }
+        this.lastAttack = hand;
+        return hand.itemStack();
+    }
+
+    @Override
+    public void setStackInHand$BetterCombat(InteractionHand handArg, ItemStack itemStack) {
+        // DUAL WIELDING LOGIC
+        // In case item got destroyed due to durability loss
+        // We empty the correct hand
+        if (comboCount < 0) {
+            // Vanilla behaviour
+            this.setItemInHand(handArg, itemStack);
+        }
+        // `handArg` argument is always `MAIN`, we can ignore it
+        net.bettercombat.api.AttackHand hand = lastAttack;
+        if (hand == null) {
+            hand = net.bettercombat.logic.PlayerAttackHelper.getCurrentAttack(this, comboCount);
+        }
+        if (hand == null) {
+            this.setItemInHand(handArg, itemStack);
+            return;
+        }
+        var redirectedHand = hand.isOffHand() ? InteractionHand.OFF_HAND : InteractionHand.MAIN_HAND;
+        this.setItemInHand(redirectedHand, itemStack);
+    }
+
+    @Override
+    public boolean disableSweeping$BetterCombat(boolean value) {
+        if (net.bettercombat.BetterCombatMod.config.server.allow_vanilla_sweeping) {
+            return value;
+        } else {
+            net.bettercombat.api.AttackHand currentHand = net.bettercombat.logic.PlayerAttackHelper.getCurrentAttack(this, this.comboCount);
+            if (currentHand != null) {
+                return false;
+            } else {
+                return value;
+            }
+        }
+    }
+    // CWS end - (BetterCombat)
 }
diff --git a/net/minecraft/server/packs/repository/ServerPacksSource.java b/net/minecraft/server/packs/repository/ServerPacksSource.java
index 3d482f3eed2d320477d5375a04ebdde22820a00f..38b45e377d719c44f69aedb58f70ad06eeec945a 100644
--- a/net/minecraft/server/packs/repository/ServerPacksSource.java
+++ b/net/minecraft/server/packs/repository/ServerPacksSource.java
@@ -48,7 +48,7 @@ public class ServerPacksSource extends BuiltInPackSource {
     public static VanillaPackResources createVanillaPackSource() {
         return new VanillaPackResourcesBuilder()
             .setMetadata(BUILT_IN_METADATA)
-            .exposeNamespace("minecraft", ResourceLocation.PAPER_NAMESPACE) // Paper
+            .exposeNamespace("minecraft", ResourceLocation.PAPER_NAMESPACE, ResourceLocation.BETTERCOMBAT_NAMESPACE) // Paper //  // CWS - (BetterCombat) Add BetterCombat ID to resource list
             .applyDevelopmentConfig()
             .pushJarResources()
             .build(VANILLA_PACK_INFO);
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 086525c31bd8167baac79a7f41e3e4d1ca783136..7b2512a3ab0bad321e2d3b0d78b8171ed152f62a 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3695,7 +3695,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return entity != null && (this == entity || this.considersEntityAsAlly(entity) || entity.considersEntityAsAlly(this));
     }
 
-    protected boolean considersEntityAsAlly(Entity entity) {
+    public boolean considersEntityAsAlly(Entity entity) { // CWS - (BetterCombat) Attack system modify
         return this.isAlliedTo(entity.getTeam());
     }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index d6be86981219f0fe0db15d36d9c85e9a56020373..e03f66c906b9124bfb203d20f886504c2912339a 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -225,7 +225,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public int deathTime;
     public float oAttackAnim;
     public float attackAnim;
-    protected int attackStrengthTicker;
+    public int attackStrengthTicker; // CWS - (BetterCombat) Attack system modify
     public final WalkAnimationState walkAnimation = new WalkAnimationState();
     public int invulnerableDuration = 20;
     public float yBodyRot;
@@ -290,6 +290,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public int shieldBlockingDelay = this.level().paperConfig().misc.shieldBlockingDelay; // Paper - Make shield blocking delay configurable
     protected boolean shouldBurnInDay = false; public boolean shouldBurnInDay() { return this.shouldBurnInDay; } public void setShouldBurnInDay(boolean shouldBurnInDay) { this.shouldBurnInDay = shouldBurnInDay; } // Purpur - API for any mob to burn daylight
     // CraftBukkit end
+    public float customKnockbackMultiplier_BetterCombat = 1; // CWS - (BetterCombat) Attack system modify
 
     protected LivingEntity(EntityType<? extends LivingEntity> entityType, Level level) {
         super(entityType, level);
@@ -2119,7 +2120,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public void knockback(double strength, double x, double z, @Nullable Entity attacker, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause eventCause) { // Paper - knockback events
         if (!canKnockback(attacker, this)) return; // Leaf - Configurable player knockback zombie
 
-        strength *= 1.0 - this.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE);
+        strength *= (1.0 - this.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE)) * customKnockbackMultiplier_BetterCombat; // CWS - (BetterCombat) Attack system modify
         if (true || !(strength <= 0.0)) { // CraftBukkit - Call event even when force is 0
             // this.hasImpulse = true; // CraftBukkit - Move down
             Vec3 deltaMovement = this.getDeltaMovement();
diff --git a/net/minecraft/world/entity/TamableAnimal.java b/net/minecraft/world/entity/TamableAnimal.java
index c9bd7212644b86be6eacdba2e12b6a19516bd2b6..7b06bff2676fd93afd81bfd24a5372d5841a3e9a 100644
--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -202,7 +202,7 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
     }
 
     @Override
-    protected boolean considersEntityAsAlly(Entity entity) {
+    public boolean considersEntityAsAlly(Entity entity) { // CWS - (BetterCombat) Attack system modify
         if (this.isTame()) {
             LivingEntity rootOwner = this.getRootOwner();
             if (entity == rootOwner) {
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index cb88b1afe261f95998a19fcb555f91143834ba3a..e8b291b0c5584428ad25125135fe84b27080949b 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -236,7 +236,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     }
 
     @Override
-    protected boolean considersEntityAsAlly(Entity entity) {
+    public boolean considersEntityAsAlly(Entity entity) { // CWS - (BetterCombat) Attack system modify
         return this.isLikedPlayer(entity) || super.considersEntityAsAlly(entity);
     }
 
diff --git a/net/minecraft/world/entity/monster/AbstractIllager.java b/net/minecraft/world/entity/monster/AbstractIllager.java
index 88075e1f49bfe7dfe1a31094a8ef64cb8abe2bf2..9d7a577cb12ba8d6671b234e1cf46b28073df0f6 100644
--- a/net/minecraft/world/entity/monster/AbstractIllager.java
+++ b/net/minecraft/world/entity/monster/AbstractIllager.java
@@ -29,7 +29,7 @@ public abstract class AbstractIllager extends Raider {
     }
 
     @Override
-    protected boolean considersEntityAsAlly(Entity entity) {
+    public boolean considersEntityAsAlly(Entity entity) { // CWS - (BetterCombat) Attack system modify
         return super.considersEntityAsAlly(entity) || entity.getType().is(EntityTypeTags.ILLAGER_FRIENDS) && this.getTeam() == null && entity.getTeam() == null;
     }
 
diff --git a/net/minecraft/world/entity/monster/Evoker.java b/net/minecraft/world/entity/monster/Evoker.java
index 5c1992a7fd5d7c5fe23ebfad35a828263d0ff93c..1a5f2cc944505e6cf57d3cf0eabf15f31cd22006 100644
--- a/net/minecraft/world/entity/monster/Evoker.java
+++ b/net/minecraft/world/entity/monster/Evoker.java
@@ -118,7 +118,7 @@ public class Evoker extends SpellcasterIllager {
     }
 
     @Override
-    protected boolean considersEntityAsAlly(Entity entity) {
+    public boolean considersEntityAsAlly(Entity entity) { // CWS - (BetterCombat) Attack system modify
         return entity == this
             || super.considersEntityAsAlly(entity)
             || entity instanceof Vex vex && vex.getOwner() != null && this.considersEntityAsAlly(vex.getOwner());
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 7d6026fdea18a7fe935973da5f7a90fd1605549a..b02357849047eb89fe4f69684eb7c9b028259cae 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1284,13 +1284,15 @@ public abstract class Player extends LivingEntity {
                         }
 
                         float f2 = f + f1;
-                        boolean flag3 = false;
+                        boolean flag3 = disableSweeping$BetterCombat(false); // CWS - (BetterCombat) Attack system modify
                         if (flag && !flag2 && !flag1 && this.onGround()) {
                             double d = this.getKnownMovement().horizontalDistanceSqr();
                             double d1 = this.getSpeed() * 2.5;
-                            if (d < Mth.square(d1) && this.getItemInHand(InteractionHand.MAIN_HAND).is(ItemTags.SWORDS)) {
-                                flag3 = true;
+                            // CWS start - (BetterCombat) Attack system modify
+                            if (d < Mth.square(d1) && this.getItemInHand(this.getHand$BetterCombat()).is(ItemTags.SWORDS)) {
+                                flag3 = disableSweeping$BetterCombat(true);
                             }
+                            // CWS end - (BetterCombat) Attack system modify
                         }
 
                         float f3 = 0.0F;
@@ -1425,11 +1427,13 @@ public abstract class Player extends LivingEntity {
                                 }
 
                                 if (weaponItem.isEmpty()) {
-                                    if (weaponItem == this.getMainHandItem()) {
-                                        this.setItemInHand(InteractionHand.MAIN_HAND, ItemStack.EMPTY);
+                                    // DomamaN202 start - Mixin PlayerEntityRangeMixin
+                                    if (weaponItem == this.getMainHandStack$BetterCombat()) {
+                                        this.setStackInHand$BetterCombat(InteractionHand.MAIN_HAND, ItemStack.EMPTY);
                                     } else {
-                                        this.setItemInHand(InteractionHand.OFF_HAND, ItemStack.EMPTY);
+                                        this.setStackInHand$BetterCombat(InteractionHand.OFF_HAND, ItemStack.EMPTY);
                                     }
+                                    // DomamaN202 end - Mixin PlayerEntityRangeMixin
                                 }
                             }
 
@@ -2448,4 +2452,22 @@ public abstract class Player extends LivingEntity {
             return this.message;
         }
     }
+
+    // CWS start - (BetterCombat)
+    public InteractionHand getHand$BetterCombat() {
+        throw new RuntimeException("Method not implemented for " + this.getClass().getName());
+    }
+
+    public ItemStack getMainHandStack$BetterCombat() {
+        throw new RuntimeException("Method not implemented for " + this.getClass().getName());
+    }
+
+    public void setStackInHand$BetterCombat(InteractionHand handArg, ItemStack itemStack) {
+        throw new RuntimeException("Method not implemented for " + this.getClass().getName());
+    }
+
+    public boolean disableSweeping$BetterCombat(boolean value) {
+        throw new RuntimeException("Method not implemented for " + this.getClass().getName());
+    }
+    // CWS end - (BetterCombat)
 }
diff --git a/net/minecraft/world/inventory/InventoryMenu.java b/net/minecraft/world/inventory/InventoryMenu.java
index 650430e225f4d8e925edd535a09e42759f8d6d5e..963eb73c6e9ee046d682967c9b8bd7268eced198 100644
--- a/net/minecraft/world/inventory/InventoryMenu.java
+++ b/net/minecraft/world/inventory/InventoryMenu.java
@@ -2,6 +2,8 @@ package net.minecraft.world.inventory;
 
 import java.util.List;
 import java.util.Map;
+
+import net.bettercombat.logic.WeaponRegistry;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.Container;
@@ -64,12 +66,33 @@ public class InventoryMenu extends AbstractCraftingMenu {
 
         this.addStandardInventorySlots(playerInventory, 8, 84);
         this.addSlot(new Slot(playerInventory, 40, 77, 62) {
+            @Override
+            public void setByPlayer(ItemStack stack) {
+                if (this.isActive()) {
+                    super.setByPlayer(stack);
+                }
+            }
+
             @Override
             public void setByPlayer(ItemStack newStack, ItemStack oldStack) {
                 owner.onEquipItem(EquipmentSlot.OFFHAND, oldStack, newStack);
                 super.setByPlayer(newStack, oldStack);
             }
 
+            @Override
+            public ItemStack getItem() {
+                if (this.isActive())
+                    return super.getItem();
+                return ItemStack.EMPTY;
+            }
+
+            @Override
+            public boolean isActive() {
+                var mainHandStack = ((Inventory) this.container).getSelectedItem();
+                var mainHandAttributes = WeaponRegistry.getAttributes(mainHandStack);
+                return mainHandAttributes == null || !mainHandAttributes.isTwoHanded();
+            }
+
             @Override
             public ResourceLocation getNoItemIcon() {
                 return InventoryMenu.EMPTY_ARMOR_SLOT_SHIELD;
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index f5ca3d6b29b11475ac56cd206464577b2d85b7dc..cfa5805ca715331c38592c891ca0122f1d95aa3d 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -94,7 +94,7 @@ import org.apache.commons.lang3.function.TriConsumer;
 import org.apache.commons.lang3.mutable.MutableBoolean;
 import org.slf4j.Logger;
 
-public final class ItemStack implements DataComponentHolder {
+public final class ItemStack implements DataComponentHolder, net.bettercombat.logic.ItemStackNBTWeaponAttributes { // CWS - (BetterCombat) Weapon Attributes
     private static final List<Component> OP_NBT_WARNING = List.of(
         Component.translatable("item.op_warning.line1").withStyle(ChatFormatting.RED, ChatFormatting.BOLD),
         Component.translatable("item.op_warning.line2").withStyle(ChatFormatting.RED),
@@ -1385,4 +1385,28 @@ public final class ItemStack implements DataComponentHolder {
     public boolean canDestroyBlock(BlockState state, Level level, BlockPos pos, Player player) {
         return this.getItem().canDestroyBlock(this, state, level, pos, player);
     }
+
+    // CWS start - (BetterCombat) Weapon Attributes
+    private boolean hasInvalidAttributes = false;
+
+    public boolean hasInvalidAttributes() {
+        return hasInvalidAttributes;
+    }
+
+    public void setInvalidAttributes(boolean invalid) {
+        this.hasInvalidAttributes = invalid;
+    }
+
+    private net.bettercombat.api.WeaponAttributes weaponAttributes;
+
+    @Override
+    public net.bettercombat.api.WeaponAttributes getWeaponAttributes() {
+        return weaponAttributes;
+    }
+
+    @Override
+    public void setWeaponAttributes(net.bettercombat.api.WeaponAttributes weaponAttributes) {
+        this.weaponAttributes = weaponAttributes;
+    }
+    // CWS end - (BetterCombat) Weapon Attributes
 }
diff --git a/net/minecraft/world/item/ProjectileWeaponItem.java b/net/minecraft/world/item/ProjectileWeaponItem.java
index d1661c9b971ba431da3f71fd6e0c3517c6625a11..a8b8b28b38dc6990cd75856a870fe498e03a22d3 100644
--- a/net/minecraft/world/item/ProjectileWeaponItem.java
+++ b/net/minecraft/world/item/ProjectileWeaponItem.java
@@ -31,12 +31,23 @@ public abstract class ProjectileWeaponItem extends Item {
     public abstract Predicate<ItemStack> getAllSupportedProjectiles();
 
     public static ItemStack getHeldProjectile(LivingEntity shooter, Predicate<ItemStack> isAmmo) {
+        // CWS start - (BetterCombat) Attack system modify
+        var originalResult = getHeldProjectile$Original(shooter, isAmmo); // Always call original first to allow others' side effects
+        if (shooter instanceof Player player) {
+            return player.getOffhandItem();
+        } else {
+            return originalResult;
+        }
+    }
+
+    public static ItemStack getHeldProjectile$Original(LivingEntity shooter, Predicate<ItemStack> isAmmo) {
         if (isAmmo.test(shooter.getItemInHand(InteractionHand.OFF_HAND))) {
             return shooter.getItemInHand(InteractionHand.OFF_HAND);
         } else {
             return isAmmo.test(shooter.getItemInHand(InteractionHand.MAIN_HAND)) ? shooter.getItemInHand(InteractionHand.MAIN_HAND) : ItemStack.EMPTY;
         }
     }
+    // CWS end - (BetterCombat) Attack system modify
 
     public abstract int getDefaultProjectileRange();
 
diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 183874d90d576d740c5d924accc5c0d7fdb8450c..15887ca274b64e25bb070ff65fab55650d058b12 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -41,6 +41,7 @@ import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EquipmentSlot;
 import net.minecraft.world.entity.EquipmentSlotGroup;
 import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.enchantment.effects.DamageImmunity;
@@ -118,6 +119,24 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
             }
         }
 
+        // CWS start - (BetterCombat) Attack system modify
+        if(entity instanceof Player player) {
+            var comboCount = ((net.bettercombat.logic.PlayerAttackProperties) player).getComboCount();
+            var currentHand = net.bettercombat.logic.PlayerAttackHelper.getCurrentAttack(player, comboCount);
+            // If striking with off-hand
+            if (currentHand != null && currentHand.isOffHand()) {
+                // Getting enchant from off-hand stack
+                if (map.get(EquipmentSlot.MAINHAND) != null) {
+                    map.remove(EquipmentSlot.MAINHAND);
+                }
+                var offHandStack = player.getOffhandItem();
+                if (!offHandStack.isEmpty()) {
+                    map.put(EquipmentSlot.OFFHAND, offHandStack);
+                }
+            }
+        }
+        // CWS end - Attack system modify
+
         return map;
     }
 
