From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 06:23:06 +0700
Subject: [PATCH] Add patch work check


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index dd6ae338fa48d52962ee0af5b1572077ba6dff91..a5b9329aa0c11e18ceb77fe022c58d0dd1422e81 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -115,6 +115,10 @@ public class Main {
             org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands = purpurConfiguration.getBoolean("settings.register-minecraft-debug-commands"); // Purpur - register minecraft debug commands
             // Purpur end - Add toggle for enchant level clamping - load config files early
 
+            // DomamaN202 start - Add patch work check
+            System.out.println("[DomamaN202] Patching really work!");
+            // DomamaN202 end - Add patch work check
+
             org.dreeam.leaf.config.LeafConfig.loadConfig(); // Leaf - Leaf config
             io.papermc.paper.plugin.PluginInitializerManager.load(optionSet); // Paper
             Bootstrap.bootStrap();
