From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 07:54:13 +0700
Subject: [PATCH] Access to Entity.considersEntityAsAlly


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 918cb68282ce179c6c3f5455f108501f7e965068..2bc827d6dc8547d9a9893e9e741f6e10229b7938 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -3696,7 +3696,9 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return entity != null && (this == entity || this.considersEntityAsAlly(entity) || entity.considersEntityAsAlly(this));
     }
 
-    protected boolean considersEntityAsAlly(Entity entity) {
+    // DomamaN202 start - Access to considersEntityAsAlly
+    public boolean considersEntityAsAlly(Entity entity) {
+    // DomamaN202 end - Access to considersEntityAsAlly
         return this.isAlliedTo(entity.getTeam());
     }
 
