From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 09:35:51 +0700
Subject: [PATCH] Access to LivingEntity.attackStrengthTicker


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 6227791579f66a11fd20748bb822c6e477ad9197..c17dd37976f26a91efd069cf14567d4839d9f483 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -225,7 +225,9 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public int deathTime;
     public float oAttackAnim;
     public float attackAnim;
+    // DomamaN202 start - Access to attackStrengthTicker
     protected int attackStrengthTicker;
+    // DomamaN202 end - Access to attackStrengthTicker
     public final WalkAnimationState walkAnimation = new WalkAnimationState();
     public int invulnerableDuration = 20;
     public float yBodyRot;
