From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 10:25:40 +0700
Subject: [PATCH] PlayerEntityMixin port


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 4dca1f9ac2ee2cd8ff4df64fe46fb8da047dbf80..80ac2a9c3682ec7b02684ae5ced14adabf5c4354 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1,9 +1,7 @@
 package net.minecraft.world.entity.player;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.Lists;
+import com.google.common.collect.*;
 import com.google.common.math.IntMath;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Either;
@@ -16,6 +14,11 @@ import java.util.OptionalInt;
 import java.util.function.Predicate;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+
+import net.bettercombat.BetterCombatMod;
+import net.bettercombat.api.AttackHand;
+import net.bettercombat.logic.PlayerAttackHelper;
+import net.bettercombat.logic.WeaponRegistry;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -74,6 +77,8 @@ import net.minecraft.world.entity.MoverType;
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.entity.SlotAccess;
 import net.minecraft.world.entity.TamableAnimal;
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.AttributeModifier;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.animal.Parrot;
@@ -224,6 +229,14 @@ public abstract class Player extends LivingEntity {
     public boolean canPortalInstant = false; // Purpur - Add portal permission bypass
     public int sixRowEnderchestSlotCount = -1; // Purpur - Barrels and enderchests 6 rows
 
+    // DomamaN202 start - Mixin PlayerEntityMixin
+    private int comboCount = 0;
+    public int getComboCount() {return comboCount;}
+    public void setComboCount(int comboCount) {this.comboCount = comboCount;}
+    private static final EntityDataAccessor<String> BETTER_COMBAT_MAIN_IDLE_ANIMATION = SynchedEntityData.defineId(Player.class, EntityDataSerializers.STRING);
+    private static final EntityDataAccessor<String> BETTER_COMBAT_OFF_IDLE_ANIMATION = SynchedEntityData.defineId(Player.class, EntityDataSerializers.STRING);
+    // DomamaN202 end - Mixin PlayerEntityMixin
+
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
@@ -310,6 +323,10 @@ public abstract class Player extends LivingEntity {
         builder.define(DATA_PLAYER_MAIN_HAND, (byte)DEFAULT_MAIN_HAND.getId());
         builder.define(DATA_SHOULDER_LEFT, new CompoundTag());
         builder.define(DATA_SHOULDER_RIGHT, new CompoundTag());
+        // DomamaN202 start - Mixin PlayerEntityMixin
+        builder.define(BETTER_COMBAT_MAIN_IDLE_ANIMATION, "");
+        builder.define(BETTER_COMBAT_OFF_IDLE_ANIMATION, "");
+        // DomamaN202 end - Mixin PlayerEntityMixin
     }
 
     @Override
@@ -418,7 +435,137 @@ public abstract class Player extends LivingEntity {
         if (this.currentImpulseContextResetGraceTime > 0) {
             this.currentImpulseContextResetGraceTime--;
         }
+
+        // DomamaN202 start - Mixin PlayerEntityMixin
+        var pose = PlayerAttackHelper.poseForPlayer(this);
+        this.entityData.set(BETTER_COMBAT_MAIN_IDLE_ANIMATION, pose.base());
+        this.entityData.set(BETTER_COMBAT_OFF_IDLE_ANIMATION, pose.offHand());
+        updateDualWieldingSpeedBoost();
+        // DomamaN202 end - Mixin PlayerEntityMixin
+    }
+
+    // DomamaN202 start - Mixin PlayerEntityMixin
+    private Multimap<Holder<Attribute>, AttributeModifier> dualWieldingAttributeMap;
+    private static final ResourceLocation dualWieldingSpeedModifierId = ResourceLocation.fromNamespaceAndPath(BetterCombatMod.ID, "dual_wield");
+
+    private void updateDualWieldingSpeedBoost() {
+        var newState = PlayerAttackHelper.isDualWielding(this);
+        var currentState = dualWieldingAttributeMap != null;
+        if (newState != currentState) {
+            if(newState) {
+                // Just started dual wielding
+                // Adding speed boost modifier
+                this.dualWieldingAttributeMap = HashMultimap.create();
+                double multiplier = BetterCombatMod.config.dual_wielding_attack_speed_multiplier - 1;
+                dualWieldingAttributeMap.put(
+                    Attributes.ATTACK_SPEED,
+                    new AttributeModifier(
+                        dualWieldingSpeedModifierId,
+                        multiplier,
+                        AttributeModifier.Operation.ADD_MULTIPLIED_BASE));
+                this.getAttributes().addTransientAttributeModifiers(dualWieldingAttributeMap);
+            } else {
+                // Just stopped dual wielding
+                // Removing speed boost modifier
+                if (dualWieldingAttributeMap != null) { // Safety first... Who knows...
+                    this.getAttributes().removeAttributeModifiers(dualWieldingAttributeMap);
+                    dualWieldingAttributeMap = null;
+                }
+            }
+        }
+    }
+
+    @Override
+    public ItemStack getItemBySlot(EquipmentSlot slot) {
+        var mainHandHasTwoHanded = false;
+        var mainHandStack = this.getMainHandItem();
+        var mainHandAttributes = WeaponRegistry.getAttributes(mainHandStack);
+        if (mainHandAttributes != null && mainHandAttributes.isTwoHanded()) {
+            mainHandHasTwoHanded = true;
+        }
+
+        var offHandHasTwoHanded = false;
+        var offHandStack = this.getOffhandItem();
+        var offHandAttributes = WeaponRegistry.getAttributes(offHandStack);
+        if(offHandAttributes != null && offHandAttributes.isTwoHanded()) {
+            offHandHasTwoHanded = true;
+        }
+
+        if (slot == EquipmentSlot.OFFHAND) {
+            if (mainHandHasTwoHanded || offHandHasTwoHanded) {
+                return (ItemStack.EMPTY);
+            }
+        }
+
+        return super.getItemBySlot(slot);
+    }
+
+    public InteractionHand getHand$BetterCombat() {
+        var currentHand = PlayerAttackHelper.getCurrentAttack(this, comboCount);
+        if (currentHand != null) {
+            return currentHand.isOffHand() ? InteractionHand.OFF_HAND : InteractionHand.MAIN_HAND;
+        } else {
+            return InteractionHand.MAIN_HAND;
+        }
+    }
+
+    private AttackHand lastAttack;
+    public ItemStack getMainHandStack$BetterCombat() {
+    // DUAL WIELDING LOGIC
+        // Here we return the off-hand stack as fake main-hand, purpose:
+        // - Getting enchants
+        // - Getting itemstack to be damaged
+        if (comboCount < 0) {
+            // Vanilla behaviour
+            return this.getMainHandItem();
+        }
+        var hand = PlayerAttackHelper.getCurrentAttack(this, comboCount);
+        if (hand == null) {
+            var isOffHand = PlayerAttackHelper.shouldAttackWithOffHand(this, comboCount);
+            if (isOffHand) {
+                return ItemStack.EMPTY;
+            } else {
+                return this.getMainHandItem();
+            }
+        }
+        this.lastAttack = hand;
+        return hand.itemStack();
+    }
+
+    public void setStackInHand$BetterCombat(InteractionHand handArg, ItemStack itemStack) {
+        // DUAL WIELDING LOGIC
+        // In case item got destroyed due to durability loss
+        // We empty the correct hand
+        if (comboCount < 0) {
+            // Vanilla behaviour
+            this.setItemInHand(handArg, itemStack);
+        }
+        // `handArg` argument is always `MAIN`, we can ignore it
+        AttackHand hand = lastAttack;
+        if (hand == null) {
+            hand = PlayerAttackHelper.getCurrentAttack(this, comboCount);
+        }
+        if (hand == null) {
+            this.setItemInHand(handArg, itemStack);
+            return;
+        }
+        var redirectedHand = hand.isOffHand() ? InteractionHand.OFF_HAND : InteractionHand.MAIN_HAND;
+        this.setItemInHand(redirectedHand, itemStack);
+    }
+
+    public boolean disableSweeping$BetterCombat(boolean value) {
+        if (BetterCombatMod.config.allow_vanilla_sweeping) {
+            return value;
+        } else {
+            AttackHand currentHand = PlayerAttackHelper.getCurrentAttack(this, this.comboCount);
+            if (currentHand != null) {
+                return false;
+            } else {
+                return value;
+            }
+        }
     }
+    // DomamaN202 end - Mixin PlayerEntityMixin
 
     @Override
     protected float getMaxHeadRotationRelativeToBody() {
@@ -1224,6 +1371,8 @@ public abstract class Player extends LivingEntity {
             );
     }
 
+    // todo: [DomamaN202] disableSweeping
+
     public void attack(Entity target) {
         // Paper start - PlayerAttackEntityEvent
         boolean willAttack = target.isAttackable() && !target.skipAttackInteraction(this); // Vanilla logic
@@ -1284,13 +1433,17 @@ public abstract class Player extends LivingEntity {
                         }
 
                         float f2 = f + f1;
-                        boolean flag3 = false;
+                        // DomamaN202 start - Mixin PlayerEntityRangeMixin
+                        boolean flag3 = disableSweeping$BetterCombat(false);
+                        // DomamaN202 end - Mixin PlayerEntityRangeMixin
                         if (flag && !flag2 && !flag1 && this.onGround()) {
                             double d = this.getKnownMovement().horizontalDistanceSqr();
                             double d1 = this.getSpeed() * 2.5;
-                            if (d < Mth.square(d1) && this.getItemInHand(InteractionHand.MAIN_HAND).is(ItemTags.SWORDS)) {
-                                flag3 = true;
+                            // DomamaN202 start - Mixin PlayerEntityRangeMixin
+                            if (d < Mth.square(d1) && this.getItemInHand(this.getHand$BetterCombat()).is(ItemTags.SWORDS)) {
+                                flag3 = disableSweeping$BetterCombat(true);
                             }
+                            // DomamaN202 end - Mixin PlayerEntityRangeMixin
                         }
 
                         float f3 = 0.0F;
@@ -1425,11 +1578,13 @@ public abstract class Player extends LivingEntity {
                                 }
 
                                 if (weaponItem.isEmpty()) {
-                                    if (weaponItem == this.getMainHandItem()) {
-                                        this.setItemInHand(InteractionHand.MAIN_HAND, ItemStack.EMPTY);
+                                    // DomamaN202 start - Mixin PlayerEntityRangeMixin
+                                    if (weaponItem == this.getMainHandStack$BetterCombat()) {
+                                        this.setStackInHand$BetterCombat(InteractionHand.MAIN_HAND, ItemStack.EMPTY);
                                     } else {
-                                        this.setItemInHand(InteractionHand.OFF_HAND, ItemStack.EMPTY);
+                                        this.setStackInHand$BetterCombat(InteractionHand.OFF_HAND, ItemStack.EMPTY);
                                     }
+                                    // DomamaN202 end - Mixin PlayerEntityRangeMixin
                                 }
                             }
 
@@ -2359,7 +2514,9 @@ public abstract class Player extends LivingEntity {
     }
 
     public double blockInteractionRange() {
-        return this.getAttributeValue(Attributes.BLOCK_INTERACTION_RANGE);
+        // DomamaN202 start - Mixin PlayerEntityRangeMixin
+        return PlayerAttackHelper.getRangeWithWeapon(this, this.getAttributeValue(Attributes.BLOCK_INTERACTION_RANGE));
+        // DomamaN202 end - Mixin PlayerEntityRangeMixin
     }
 
     public double entityInteractionRange() {
