From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 10:29:57 +0700
Subject: [PATCH] EnchantmentMixin port


diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 183874d90d576d740c5d924accc5c0d7fdb8450c..65f5aa3175fa97fafc1641c477808eef6654864e 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -12,6 +12,9 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 import java.util.function.Consumer;
+
+import net.bettercombat.logic.PlayerAttackHelper;
+import net.bettercombat.logic.PlayerAttackProperties;
 import net.minecraft.ChatFormatting;
 import net.minecraft.Util;
 import net.minecraft.core.Holder;
@@ -41,6 +44,7 @@ import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EquipmentSlot;
 import net.minecraft.world.entity.EquipmentSlotGroup;
 import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.enchantment.effects.DamageImmunity;
@@ -118,6 +122,24 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
             }
         }
 
+        // DomamaN202 start - Mixin EnchantmentMixin
+        if(entity instanceof Player player) {
+            var comboCount = ((PlayerAttackProperties) player).getComboCount();
+            var currentHand = PlayerAttackHelper.getCurrentAttack(player, comboCount);
+            // If striking with off-hand
+            if (currentHand != null && currentHand.isOffHand()) {
+                // Getting enchant from off-hand stack
+                if (map.get(EquipmentSlot.MAINHAND) != null) {
+                    map.remove(EquipmentSlot.MAINHAND);
+                }
+                var offHandStack = player.getOffhandItem();
+                if (!offHandStack.isEmpty()) {
+                    map.put(EquipmentSlot.OFFHAND, offHandStack);
+                }
+            }
+        }
+        // DomamaN202 end - Mixin EnchantmentMixin
+
         return map;
     }
 
