From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 10:36:53 +0700
Subject: [PATCH] RangedWeaponItemMixin port


diff --git a/net/minecraft/world/item/ProjectileWeaponItem.java b/net/minecraft/world/item/ProjectileWeaponItem.java
index f37fd3b9ab725e5b8eb7fccf9b35bbc09e3a2d0c..b48bd069d45c077ee40611cfafc77246baddb031 100644
--- a/net/minecraft/world/item/ProjectileWeaponItem.java
+++ b/net/minecraft/world/item/ProjectileWeaponItem.java
@@ -30,13 +30,25 @@ public abstract class ProjectileWeaponItem extends Item {
 
     public abstract Predicate<ItemStack> getAllSupportedProjectiles();
 
+
+    // DomamaN202 start - Mixin RangedWeaponItemMixin
     public static ItemStack getHeldProjectile(LivingEntity shooter, Predicate<ItemStack> isAmmo) {
+        var originalResult = getHeldProjectile$Original(shooter, isAmmo); // Always call original first to allow others' side effects
+        if (shooter instanceof Player player) {
+            return player.getOffhandItem();
+        } else {
+            return originalResult;
+        }
+    }
+
+    public static ItemStack getHeldProjectile$Original(LivingEntity shooter, Predicate<ItemStack> isAmmo) {
         if (isAmmo.test(shooter.getItemInHand(InteractionHand.OFF_HAND))) {
             return shooter.getItemInHand(InteractionHand.OFF_HAND);
         } else {
             return isAmmo.test(shooter.getItemInHand(InteractionHand.MAIN_HAND)) ? shooter.getItemInHand(InteractionHand.MAIN_HAND) : ItemStack.EMPTY;
         }
     }
+    // DomamaN202 end - Mixin RangedWeaponItemMixin
 
     public abstract int getDefaultProjectileRange();
 
