From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 10:54:52 +0700
Subject: [PATCH] Access to Entity.considersEntityAsAlly


diff --git a/net/minecraft/world/entity/monster/AbstractIllager.java b/net/minecraft/world/entity/monster/AbstractIllager.java
index 88075e1f49bfe7dfe1a31094a8ef64cb8abe2bf2..f25923155410b0263aede7f948006734c9a91f06 100644
--- a/net/minecraft/world/entity/monster/AbstractIllager.java
+++ b/net/minecraft/world/entity/monster/AbstractIllager.java
@@ -29,7 +29,9 @@ public abstract class AbstractIllager extends Raider {
     }
 
     @Override
-    protected boolean considersEntityAsAlly(Entity entity) {
+    // DomamaN202 start - Access to considersEntityAsAlly
+    public boolean considersEntityAsAlly(Entity entity) {
+    // DomamaN202 end - Access to considersEntityAsAlly
         return super.considersEntityAsAlly(entity) || entity.getType().is(EntityTypeTags.ILLAGER_FRIENDS) && this.getTeam() == null && entity.getTeam() == null;
     }
 
