From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 10:55:12 +0700
Subject: [PATCH] Access to Entity.considersEntityAsAlly


diff --git a/net/minecraft/world/entity/monster/Evoker.java b/net/minecraft/world/entity/monster/Evoker.java
index 5c1992a7fd5d7c5fe23ebfad35a828263d0ff93c..0c74a67d1c4d62b079ac993b2f5bba2551dce045 100644
--- a/net/minecraft/world/entity/monster/Evoker.java
+++ b/net/minecraft/world/entity/monster/Evoker.java
@@ -118,7 +118,9 @@ public class Evoker extends SpellcasterIllager {
     }
 
     @Override
-    protected boolean considersEntityAsAlly(Entity entity) {
+    // DomamaN202 start - Access to considersEntityAsAlly
+    public boolean considersEntityAsAlly(Entity entity) {
+    // DomamaN202 end - Access to considersEntityAsAlly
         return entity == this
             || super.considersEntityAsAlly(entity)
             || entity instanceof Vex vex && vex.getOwner() != null && this.considersEntityAsAlly(vex.getOwner());
