From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 17:22:18 +0700
Subject: [PATCH] BetterCombat data loading


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 9eb35364757b5748f7228c1557e9350edc805440..8c5ad14170134fe0d6bd20bb8d2b757c720156df 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -50,6 +50,8 @@ import java.util.function.Function;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
 import javax.imageio.ImageIO;
+
+import net.bettercombat.BetterCombatMod;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.FileUtil;
@@ -473,6 +475,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // CraftBukkit end
         this.paperConfigurations = services.paperConfigurations(); // Paper - add paper configuration files
         this.galeConfigurations = services.galeConfigurations(); // Gale - Gale configuration
+        // DomamaN202 start - Load BetterCombat data
+        BetterCombatMod.loadWeaponAttributes(this);
+        // DomamaN202 end - Load BetterCombat data
     }
 
     private void readScoreboard(DimensionDataStorage dataStorage) {
