From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Tue, 8 Jul 2025 18:05:37 +0700
Subject: [PATCH] BetterCombat resources


diff --git a/net/minecraft/resources/ResourceLocation.java b/net/minecraft/resources/ResourceLocation.java
index 98e27b3cea8ee8c0ba3445a3153bc228c104f3df..e7f63c6e58fb1f0d68af313a8ac66c06ad1f8e87 100644
--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -24,6 +24,7 @@ public final class ResourceLocation implements Comparable<ResourceLocation> {
     public static final String DEFAULT_NAMESPACE = "minecraft";
     public static final String REALMS_NAMESPACE = "realms";
     public static final String PAPER_NAMESPACE = "paper"; // Paper
+    public static final String BETTERCOMBAT_NAMESPACE = "bettercombat";
     private final String namespace;
     private final String path;
 
diff --git a/net/minecraft/server/packs/repository/ServerPacksSource.java b/net/minecraft/server/packs/repository/ServerPacksSource.java
index 3d482f3eed2d320477d5375a04ebdde22820a00f..0df4996590b3eb5b244a3c93564c2a3606d29a29 100644
--- a/net/minecraft/server/packs/repository/ServerPacksSource.java
+++ b/net/minecraft/server/packs/repository/ServerPacksSource.java
@@ -48,7 +48,7 @@ public class ServerPacksSource extends BuiltInPackSource {
     public static VanillaPackResources createVanillaPackSource() {
         return new VanillaPackResourcesBuilder()
             .setMetadata(BUILT_IN_METADATA)
-            .exposeNamespace("minecraft", ResourceLocation.PAPER_NAMESPACE) // Paper
+            .exposeNamespace("minecraft", ResourceLocation.PAPER_NAMESPACE, ResourceLocation.BETTERCOMBAT_NAMESPACE) // Paper // DomamaN202
             .applyDevelopmentConfig()
             .pushJarResources()
             .build(VANILLA_PACK_INFO);
