From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Wed, 9 Jul 2025 09:14:12 +0700
Subject: [PATCH] Naming fix


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 0f61eb631b1f9c5a83ed3c654eb2cff2978eef5f..9468a7696239b9cb00805540fd2186849409bbcc 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -456,7 +456,7 @@ public abstract class Player extends LivingEntity {
                 // Just started dual wielding
                 // Adding speed boost modifier
                 this.dualWieldingAttributeMap = HashMultimap.create();
-                double multiplier = BetterCombatMod.config.dual_wielding_attack_speed_multiplier - 1;
+                double multiplier = BetterCombatMod.config.server.dual_wielding_attack_speed_multiplier - 1;
                 dualWieldingAttributeMap.put(
                     Attributes.ATTACK_SPEED,
                     new AttributeModifier(
@@ -554,7 +554,7 @@ public abstract class Player extends LivingEntity {
     }
 
     public boolean disableSweeping$BetterCombat(boolean value) {
-        if (BetterCombatMod.config.allow_vanilla_sweeping) {
+        if (BetterCombatMod.config.server.allow_vanilla_sweeping) {
             return value;
         } else {
             AttackHand currentHand = PlayerAttackHelper.getCurrentAttack(this, this.comboCount);
