From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Wed, 9 Jul 2025 12:00:28 +0700
Subject: [PATCH] ServerConfigurationConnectionEvents impl


diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index 25c6e9f1b88ea7bfa2c7f1d130d56c581202960e..cc470d40935b6faa23add6de197ad2e45fdba697 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -90,6 +90,7 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
     }
 
     public void startConfiguration() {
+        net.fabricmc.fabric.api.networking.v1.ServerConfigurationConnectionEvents.CONFIGURE.invoker().onSendConfiguration(this, server); // DomamaN202 - ServerConfigurationConnectionEvents
         new io.papermc.paper.event.connection.configuration.PlayerConnectionInitialConfigureEvent(this.paperConnection).callEvent(); // Paper
         this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
         ServerLinks serverLinks = this.server.serverLinks();
