From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: DomamaN202 <vip.domaman@mail.ru>
Date: Wed, 9 Jul 2025 13:12:54 +0700
Subject: [PATCH] DmNServerPlayNetworking impl


diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 0c1fa0525819a1a1a78730d08a2592975fafbec5..4ad070748d00dfd3c99c5f141ce318f88856248b 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -217,22 +217,6 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
             }
 
             this.cserver.getMessenger().dispatchIncomingMessage(paperConnection(), identifier.toString(), data);
-
-            // DomamaN202 start - DmNServerPlayNetworking
-            LOGGER.warn("Custom payload: " + packet.payload().type().id());
-            var play_handler = DmNServerPlayNetworking.PLAY_HANDLERS.get(packet.payload().type().id());
-            LOGGER.warn("Custom handler A: " + play_handler);
-            if (play_handler != null) {
-                ((DmNServerPlayNetworking.PlayPayloadHandler<CustomPacketPayload>) play_handler).receive(packet.payload(), this.connection.getPlayer());
-                return;
-            }
-            var config_handler = DmNServerPlayNetworking.CONFIG_HANDLERS.get(packet.payload().type().id());
-            LOGGER.warn("Custom handler B: " + config_handler);
-            if (config_handler != null && this instanceof ServerConfigurationPacketListenerImpl) {
-                ((DmNServerPlayNetworking.ConfigPayloadHandler<CustomPacketPayload>) config_handler).receive(packet.payload(), (ServerConfigurationPacketListenerImpl) this, this.connection.getPlayer());
-                return;
-            }
-            // DomamaN202 end - DmNServerPlayNetworking
         } catch (final Exception e) {
             ServerGamePacketListenerImpl.LOGGER.error("Couldn't handle custom payload on channel {}", identifier, e);
             this.disconnect(net.minecraft.network.chat.Component.literal("Invalid custom payload payload!"), io.papermc.paper.connection.DisconnectionReason.INVALID_PAYLOAD); // Paper - kick event cause
diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index 4159acbed71318a05a0e6165e6a97f078beadb68..636f01bcf7577e700f16129838b3b6634e099902 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -19,6 +19,7 @@ import net.minecraft.network.protocol.common.ClientboundServerLinksPacket;
 import net.minecraft.network.protocol.common.ServerboundClientInformationPacket;
 import net.minecraft.network.protocol.common.ServerboundResourcePackPacket;
 import net.minecraft.network.protocol.common.custom.BrandPayload;
+import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
 import net.minecraft.network.protocol.configuration.ClientboundUpdateEnabledFeaturesPacket;
 import net.minecraft.network.protocol.configuration.ServerConfigurationPacketListener;
 import net.minecraft.network.protocol.configuration.ServerboundFinishConfigurationPacket;
@@ -36,6 +37,7 @@ import net.minecraft.server.packs.repository.KnownPack;
 import net.minecraft.server.players.PlayerList;
 import net.minecraft.world.flag.FeatureFlags;
 import org.slf4j.Logger;
+import ru.cws.network.DmNServerPlayNetworking;
 
 public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketListenerImpl implements ServerConfigurationPacketListener, TickablePacketListener {
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -275,6 +277,21 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
 
     @Override
     public void handleCustomPayload(net.minecraft.network.protocol.common.ServerboundCustomPayloadPacket packet) {
+        // DomamaN202 start - DmNServerPlayNetworking
+        LOGGER.warn("Custom payload: " + packet.payload().type().id());
+        var play_handler = DmNServerPlayNetworking.PLAY_HANDLERS.get(packet.payload().type().id());
+        LOGGER.warn("Custom handler A: " + play_handler);
+        if (play_handler != null) {
+            ((DmNServerPlayNetworking.PlayPayloadHandler<CustomPacketPayload>) play_handler).receive(packet.payload(), this.connection.getPlayer());
+            return;
+        }
+        var config_handler = DmNServerPlayNetworking.CONFIG_HANDLERS.get(packet.payload().type().id());
+        LOGGER.warn("Custom handler B: " + config_handler);
+        if (config_handler != null) {
+            ((DmNServerPlayNetworking.ConfigPayloadHandler<CustomPacketPayload>) config_handler).receive(packet.payload(), this, this.connection.getPlayer());
+            return;
+        }
+        // DomamaN202 end - DmNServerPlayNetworking
         super.handleCustomPayload(packet);
     }
 
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index caf2ddab06eb45c150b7f8119b238d013aae6d10..4bb52c25e7d5c8eb5ec299b49ad1804c20547be1 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3798,6 +3798,15 @@ public class ServerGamePacketListenerImpl
 
     @Override
     public void handleCustomPayload(ServerboundCustomPayloadPacket packet) {
+        // DomamaN202 start - DmNServerPlayNetworking
+        LOGGER.warn("Custom payload: " + packet.payload().type().id());
+        var play_handler = DmNServerPlayNetworking.PLAY_HANDLERS.get(packet.payload().type().id());
+        LOGGER.warn("Custom handler A: " + play_handler);
+        if (play_handler != null) {
+            ((DmNServerPlayNetworking.PlayPayloadHandler<CustomPacketPayload>) play_handler).receive(packet.payload(), this.connection.getPlayer());
+            return;
+        }
+        // DomamaN202 end - DmNServerPlayNetworking
         // Leaves start - protocol
         if (packet.payload() instanceof org.leavesmc.leaves.protocol.core.LeavesCustomPayload leavesPayload) {
             org.leavesmc.leaves.protocol.core.LeavesProtocolManager.handlePayload(player, leavesPayload);
